// nodame/lib/notification.js
// Generated by CoffeeScript 1.9.3
(function() {
  var Notification, Slack;

  Slack = require('slack-client');

  Notification = (function() {
    function Notification() {
      var auto_mark, auto_reconnect;
      this.NOTIFICATION = nodame.config('notification');
      this.CLIENTS = this.NOTIFICATION.clients;
      if (this.NOTIFICATION.enable) {
        if (this.CLIENTS.slack.enable) {
          this.token = this.CLIENTS.slack.token;
          this.channel_id = this.CLIENTS.slack.channel_id;
          if (!((this.token != null) || (this.channel_id != null))) {
            return;
          }
          auto_reconnect = this.CLIENTS.slack.auto_reconnect;
          auto_mark = this.CLIENTS.slack.auto_mark;
          this.slack = new Slack(this.token, auto_reconnect, auto_mark);
          this.slack.login();
        }
      }
      return;
    }

    Notification.prototype.send = function(title, message) {
      var channel_id, name, ref, ref1, url;
      if (title == null) {
        title = 'notification';
      }
      if (message == null) {
        message = '';
      }
      if (this.NOTIFICATION.enable) {
        if (this.CLIENTS.slack.enable) {
          if (message != null) {
            message = "\n\n" + message;
          }
          name = (ref = this.CLIENTS.slack.name) != null ? ref : nodame.config('app.name');
          url = (ref1 = this.CLIENTS.slack.url) != null ? ref1 : nodame.config('url.base');
          if (title == null) {
            title = 'notification';
          }
          channel_id = this.channel_id;
          this.slack.on('open', function() {
            var body, channel, env, group, groups, id, time;
            groups = (function() {
              var ref2, results;
              ref2 = this.groups;
              results = [];
              for (id in ref2) {
                group = ref2[id];
                if (group.is_open && !group.is_archived) {
                  results.push(group.name);
                }
              }
              return results;
            }).call(this);
            channel = this.getChannelGroupOrDMByID(channel_id);
            time = new Date();
            env = nodame.env();
            body = ">>>*[" + name + " (" + env + ")]* _" + title + "_ \n_" + time + "_\n" + url + message;
            channel.send(body);
            return console.log('Notification sent to slack.');
          });
        }
      }
    };

    return Notification;

  })();

  module.exports = Notification;

}).call(this);
